### PR TITLE
feat(webhooks): allow manually retrying delivery of outgoing webhooks

### PR DESCRIPTION
**Pull request by SanchithHegde**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR includes changes in the outgoing webhooks flow to allow manually retrying webhook deliveries, and adds an API endpoint for the same.

- URL: `/events/{merchant_id_or_profile_id}/{event_id}/retry`
- Authentication: Admin API authentication + JWT
- The endpoint must return the request and response information for the attempted retry.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

This PR adds a new endpoint to retry outgoing webhook delivery.

## Motivation and Context

This allows users to debug issues in case the delivery of webhooks to their application server fails, without having to wait for the application to automatically retry at set intervals.

## How did you test it?

This PR adds a new endpoint and testing it is pretty straightforward. It can be tried out using a request like so:

```shell
curl -X POST -H 'api-key: {{admin_api_key}}' '{{baseUrl}}/events/{{merchant_id_or_profile_id}}/{{event_id}}/retry'
```

- In case of errors hitting the merchant server (say URL is inaccessible):

  ```shell
  curl -X POST -H 'api-key: test_admin' 'http://localhost:8080/events/pro_H7FbvX2pBuJLjlPhkpRa/evt_018e62ea2215731c915c3a47e4753e4a/retry' 2>/dev/null | jq '.'
  {
    "event_id": "evt_018e6581512d7b9bba60ec4ca44f364f",
    "merchant_id": "merchant_1710609960",
    "profile_id": "pro_H7FbvX2pBuJLjlPhkpRa",
    "object_id": "pay_6AMmb6rtlPDarBLlotAH",
    "event_type": "payment_succeeded",
    "event_class": "payments",
    "is_delivery_successful": false,
    "initial_attempt_id": "evt_018e48d5ad6c73d5a289342e6132b224",
    "created": "2024-03-22T09:32:37.550Z",
    "request": {
      "body": "{\"merchant_id\":\"merchant_1710609960\",\"event_id\":\"evt_018e48d5ad6c73d5a289342e6132b224\",\"event_type\":\"payment_succeeded\",\"content\":{\"type\":\"payment_details\",\"object\":{\"payment_id\":\"pay_6AMmb6rtlPDarBLlotAH\",\"merchant_id\":\"merchant_1710609960\",\"status\":\"succeeded\",\"amount\":6540,\"net_amount\":6540,\"amount_capturable\":0,\"amount_received\":6540,\"connector\":\"stripe\",\"client_secret\":\"pay_6AMmb6rtlPDarBLlotAH_secret_1hjI8KVOKf5xhMXBTwRj\",\"created\":\"2024-03-16T19:55:45.335Z\",\"currency\":\"USD\",\"customer_id\":\"StripeCustomer\",\"description\":\"Its my first payment request\",\"refunds\":null,\"disputes\":null,\"mandate_id\":null,\"mandate_data\":null,\"setup_future_usage\":null,\"off_session\":null,\"capture_on\":null,\"capture_method\":\"automatic\",\"payment_method\":\"card\",\"payment_method_data\":{\"card\":{\"last4\":\"4242\",\"card_type\":null,\"card_network\":null,\"card_issuer\":null,\"card_issuing_country\":null,\"card_isin\":\"424242\",\"card_extended_bin\":\"42424242\",\"card_exp_month\":\"10\",\"card_exp_year\":\"25\",\"card_holder_name\":\"joseph Doe\"},\"billing\":{\"address\":{\"city\":\"San Fransico\",\"country\":\"US\",\"line1\":\"1467\",\"line2\":\"Harrison Street\",\"line3\":\"Harrison Street\",\"zip\":\"94122\",\"state\":\"California\",\"first_name\":\"joseph\",\"last_name\":\"Doe\"},\"phone\":{\"number\":\"8056594427\",\"country_code\":\"+91\"},\"email\":null}},\"payment_token\":null,\"shipping\":{\"address\":{\"city\":\"San Fransico\",\"country\":\"US\",\"line1\":\"1467\",\"line2\":\"Harrison Street\",\"line3\":\"Harrison Street\",\"zip\":\"94122\",\"state\":\"California\",\"first_name\":\"joseph\",\"last_name\":\"Doe\"},\"phone\":{\"number\":\"8056594427\",\"country_code\":\"+91\"},\"email\":null},\"billing\":{\"address\":{\"city\":\"San Fransico\",\"country\":\"US\",\"line1\":\"1467\",\"line2\":\"Harrison Street\",\"line3\":\"Harrison Street\",\"zip\":\"94122\",\"state\":\"California\",\"first_name\":\"joseph\",\"last_name\":\"Doe\"},\"phone\":{\"number\":\"8056594427\",\"country_code\":\"+91\"},\"email\":null},\"order_details\":null,\"email\":\"guest@example.com\",\"name\":\"John Doe\",\"phone\":\"999999999\",\"return_url\":\"https://google.com/\",\"authentication_type\":\"no_three_ds\",\"statement_descriptor_name\":\"joseph\",\"statement_descriptor_suffix\":\"JS\",\"next_action\":null,\"cancellation_reason\":null,\"error_code\":null,\"error_message\":null,\"unified_code\":null,\"unified_message\":null,\"payment_experience\":null,\"payment_method_type\":\"credit\",\"connector_label\":null,\"business_country\":null,\"business_label\":\"default\",\"business_sub_label\":null,\"allowed_payment_method_types\":null,\"ephemeral_key\":{\"customer_id\":\"StripeCustomer\",\"created_at\":1710618945,\"expires\":1710622545,\"secret\":\"epk_5736317c91e04787838bd300d6f53877\"},\"manual_retry_allowed\":false,\"connector_transaction_id\":\"pi_3Ov3KXD5R7gDAGff1ZbtdptB\",\"frm_message\":null,\"metadata\":{\"udf1\":\"value1\",\"login_date\":\"2019-09-10T10:11:12Z\",\"new_customer\":\"true\"},\"connector_metadata\":null,\"feature_metadata\":null,\"reference_id\":\"pi_3Ov3KXD5R7gDAGff1ZbtdptB\",\"payment_link\":null,\"profile_id\":\"pro_H7FbvX2pBuJLjlPhkpRa\",\"surcharge_details\":null,\"attempt_count\":1,\"merchant_decision\":null,\"merchant_connector_id\":\"mca_qeS5H2fLkr8GLzN9SNWd\",\"incremental_authorization_allowed\":null,\"authorization_count\":null,\"incremental_authorizations\":null,\"external_authentication_details\":null,\"external_3ds_authentication_attempted\":false,\"expires_on\":\"2024-03-16T20:10:45.335Z\",\"fingerprint\":null,\"payment_method_id\":null,\"payment_method_status\":null}},\"timestamp\":\"2024-03-16T19:55:46.924Z\"}",
      "headers": [
        [
          "content-type",
          "application/json"
        ],
        [
          "X-Webhook-Signature-512",
          "68dd7abf747b5c33eaef5d7450990594aff5b260e79a33405539311f037efc2b70f86b062bcab9a231bd8fa8e8141df078e5834d52add7751da48247bf31ece5"
        ]
      ]
    },
    "response": {
      "body": null,
      "headers": null,
      "status_code": null,
      "error_message": "Unable to send request to merchant server"
    },
    "delivery_attempt": "manual_retry"
  }
  ```

  Note:

    - `initial_attempt_id` would not contain the `event_id` of the event being retried, but the `initial_attempt_id` of the event being retried. This would allow us to list all attempts including manual delivery attempts related to the initial attempt.
    - `delivery_attempt` is `manual_retry` in this case
    - `response.error_message` is non-null

- Successful delivery:

  ```shell
  curl -X POST -H 'api-key: test_admin' 'http://localhost:8080/events/pro_H7FbvX2pBuJLjlPhkpRa/evt_018e62ea2215731c915c3a47e4753e4a/retry' 2>/dev/null | jq '.'
  {
    "event_id": "evt_018e6589a52e75a9bb0dbb72fd6b6434",
    "merchant_id": "merchant_1710609960",
    "profile_id": "pro_H7FbvX2pBuJLjlPhkpRa",
    "object_id": "pay_6AMmb6rtlPDarBLlotAH",
    "event_type": "payment_succeeded",
    "event_class": "payments",
    "is_delivery_successful": true,
    "initial_attempt_id": "evt_018e48d5ad6c73d5a289342e6132b224",
    "created": "2024-03-22T09:41:43.342Z",
    "request": {
      "body": "{\"merchant_id\":\"merchant_1710609960\",\"event_id\":\"evt_018e48d5ad6c73d5a289342e6132b224\",\"event_type\":\"payment_succeeded\",\"content\":{\"type\":\"payment_details\",\"object\":{\"payment_id\":\"pay_6AMmb6rtlPDarBLlotAH\",\"merchant_id\":\"merchant_1710609960\",\"status\":\"succeeded\",\"amount\":6540,\"net_amount\":6540,\"amount_capturable\":0,\"amount_received\":6540,\"connector\":\"stripe\",\"client_secret\":\"pay_6AMmb6rtlPDarBLlotAH_secret_1hjI8KVOKf5xhMXBTwRj\",\"created\":\"2024-03-16T19:55:45.335Z\",\"currency\":\"USD\",\"customer_id\":\"StripeCustomer\",\"description\":\"Its my first payment request\",\"refunds\":null,\"disputes\":null,\"mandate_id\":null,\"mandate_data\":null,\"setup_future_usage\":null,\"off_session\":null,\"capture_on\":null,\"capture_method\":\"automatic\",\"payment_method\":\"card\",\"payment_method_data\":{\"card\":{\"last4\":\"4242\",\"card_type\":null,\"card_network\":null,\"card_issuer\":null,\"card_issuing_country\":null,\"card_isin\":\"424242\",\"card_extended_bin\":\"42424242\",\"card_exp_month\":\"10\",\"card_exp_year\":\"25\",\"card_holder_name\":\"joseph Doe\"},\"billing\":{\"address\":{\"city\":\"San Fransico\",\"country\":\"US\",\"line1\":\"1467\",\"line2\":\"Harrison Street\",\"line3\":\"Harrison Street\",\"zip\":\"94122\",\"state\":\"California\",\"first_name\":\"joseph\",\"last_name\":\"Doe\"},\"phone\":{\"number\":\"8056594427\",\"country_code\":\"+91\"},\"email\":null}},\"payment_token\":null,\"shipping\":{\"address\":{\"city\":\"San Fransico\",\"country\":\"US\",\"line1\":\"1467\",\"line2\":\"Harrison Street\",\"line3\":\"Harrison Street\",\"zip\":\"94122\",\"state\":\"California\",\"first_name\":\"joseph\",\"last_name\":\"Doe\"},\"phone\":{\"number\":\"8056594427\",\"country_code\":\"+91\"},\"email\":null},\"billing\":{\"address\":{\"city\":\"San Fransico\",\"country\":\"US\",\"line1\":\"1467\",\"line2\":\"Harrison Street\",\"line3\":\"Harrison Street\",\"zip\":\"94122\",\"state\":\"California\",\"first_name\":\"joseph\",\"last_name\":\"Doe\"},\"phone\":{\"number\":\"8056594427\",\"country_code\":\"+91\"},\"email\":null},\"order_details\":null,\"email\":\"guest@example.com\",\"name\":\"John Doe\",\"phone\":\"999999999\",\"return_url\":\"https://google.com/\",\"authentication_type\":\"no_three_ds\",\"statement_descriptor_name\":\"joseph\",\"statement_descriptor_suffix\":\"JS\",\"next_action\":null,\"cancellation_reason\":null,\"error_code\":null,\"error_message\":null,\"unified_code\":null,\"unified_message\":null,\"payment_experience\":null,\"payment_method_type\":\"credit\",\"connector_label\":null,\"business_country\":null,\"business_label\":\"default\",\"business_sub_label\":null,\"allowed_payment_method_types\":null,\"ephemeral_key\":{\"customer_id\":\"StripeCustomer\",\"created_at\":1710618945,\"expires\":1710622545,\"secret\":\"epk_5736317c91e04787838bd300d6f53877\"},\"manual_retry_allowed\":false,\"connector_transaction_id\":\"pi_3Ov3KXD5R7gDAGff1ZbtdptB\",\"frm_message\":null,\"metadata\":{\"udf1\":\"value1\",\"login_date\":\"2019-09-10T10:11:12Z\",\"new_customer\":\"true\"},\"connector_metadata\":null,\"feature_metadata\":null,\"reference_id\":\"pi_3Ov3KXD5R7gDAGff1ZbtdptB\",\"payment_link\":null,\"profile_id\":\"pro_H7FbvX2pBuJLjlPhkpRa\",\"surcharge_details\":null,\"attempt_count\":1,\"merchant_decision\":null,\"merchant_connector_id\":\"mca_qeS5H2fLkr8GLzN9SNWd\",\"incremental_authorization_allowed\":null,\"authorization_count\":null,\"incremental_authorizations\":null,\"external_authentication_details\":null,\"external_3ds_authentication_attempted\":false,\"expires_on\":\"2024-03-16T20:10:45.335Z\",\"fingerprint\":null,\"payment_method_id\":null,\"payment_method_status\":null}},\"timestamp\":\"2024-03-16T19:55:46.924Z\"}",
      "headers": [
        [
          "content-type",
          "application/json"
        ],
        [
          "X-Webhook-Signature-512",
          "68dd7abf747b5c33eaef5d7450990594aff5b260e79a33405539311f037efc2b70f86b062bcab9a231bd8fa8e8141df078e5834d52add7751da48247bf31ece5"
        ]
      ]
    },
    "response": {
      "body": "",
      "headers": [
        [
          "content-length",
          "0"
        ],
        [
          "date",
          "Fri, 22 Mar 2024 09:41:44 GMT"
        ],
        [
          "access-control-allow-origin",
          "*"
        ],
        [
          "set-cookie",
          "ARRAffinity=669b1a143e31fb4dd62d7cac87cd552f46ae1eaa9e19a1bc7e87014967e06b05;Path=/;HttpOnly;Secure;Domain=smee.io"
        ],
        [
          "set-cookie",
          "ARRAffinitySameSite=669b1a143e31fb4dd62d7cac87cd552f46ae1eaa9e19a1bc7e87014967e06b05;Path=/;HttpOnly;SameSite=None;Secure;Domain=smee.io"
        ],
        [
          "request-context",
          "appId=cid-v1:"
        ],
        [
          "x-powered-by",
          "Express"
        ]
      ],
      "status_code": 200,
      "error_message": null
    },
    "delivery_attempt": "manual_retry"
  }
  ```

- The above request should behave the same way by replacing the profile ID with the merchant ID, or replacing the `api-key` header and value with the JWT token received after user sign in.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by ThisIsMani — DISMISSED** [review]
Auth specific changes looks fine.

File: crates/router/src/core/webhooks/webhook_events.rs [hunk 1]
```diff
@@ -153,10 +155,108 @@ pub async fn list_delivery_attempts(
...
+    let event_to_retry = store
+        .find_event_by_merchant_id_event_id(&key_store.merchant_id, &event_id, &key_store)
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::EventNotFound)?;
+
+    let business_profile = match account {
+        MerchantAccountOrBusinessProfile::MerchantAccount(_) => {
+            let business_profile_id = event_to_retry
+                .business_profile_id
+                .get_required_value("business_profile_id")
```
Comment (hrithikesh026) [inline crates/router/src/core/webhooks/webhook_events.rs:178]: Shouldn't we check for profile_id in `merchant_account.default_profile_id` ?

File: crates/router/src/core/webhooks/webhook_events.rs [hunk 2]
```diff
@@ -166,10 +266,17 @@ async fn determine_identifier_and_get_key_store(
         .await
     {
         // Valid merchant ID
-        Ok(key_store) => Ok((
-            MerchantIdOrProfileId::MerchantId(merchant_id_or_profile_id),
-            key_store,
-        )),
+        Ok(key_store) => {
+            let merchant_account = store
+                .find_merchant_account_by_merchant_id(&merchant_id_or_profile_id, &key_store)
```
Comment (hrithikesh026) [inline crates/router/src/core/webhooks/webhook_events.rs:271]: passing `merchant_id_or_profile_id` as an argument to this function is a bit confusing. You can get merchant_id from key_store.

File: openapi/openapi_spec.json [hunk 3]
```diff
@@ -11677,14 +11673,22 @@
...
             "minimum": 0
+          },
+          "error_message": {
+            "type": "string",
+            "description": "Error message in case any error occurred when trying to deliver the webhook.",
+            "example": 200,
+            "nullable": true
           }
         }
       },
```
Comment (hrithikesh026) [inline openapi/openapi_spec.json:11694]: openapi documentation is not generated for the new endpoint.

Comment (sahkal) [reply openapi/openapi_spec.json:11694]: +1

**Review by sahkal — COMMENTED** [review]
regenerate open-api specs other than that LGTM!

File: crates/router/src/core/webhooks/webhook_events.rs (hunk 1 above)
Comment (SanchithHegde) [reply crates/router/src/core/webhooks/webhook_events.rs:178]: If the `business_profile_id` is not present in the event, then we're trying to retry a very old event which does not have many of the fields populated, which I don't want to support. We also need the `request` field to be populated, to be able to read the request body and headers sent previously and retry the webhook.

Moreover, if there are multiple business profiles associated with the merchant account and the default profile ID is `profile1`, but this event was related to `profile2`, then we'd end up populating incorrect data by populating `business_profile_id` as `profile1` in the database.

File: openapi/openapi_spec.json (hunk 3 above)
Comment (SanchithHegde) [reply openapi/openapi_spec.json:11760]: Done, thanks for pointing this out!

File: crates/router/src/core/webhooks/webhook_events.rs (hunk 2 above)
Comment (SanchithHegde) [reply crates/router/src/core/webhooks/webhook_events.rs:273]: Correct. At this point, `merchant_id_or_profile_id` and `key_store.merchant_id` would have the same value.
I've added a comment to clarify the usage of `merchant_id_or_profile_id`

**Review by hrithikesh026 — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]
LGTM

**Pull request merged**